server/chat-commands/room-settings.ts

```diff
@@ -17,218 +17,32 @@ const SLOWCHAT_USER_REQUIREMENT = 10;
-		if (this.room.blackjackDisabled) {
-			return `${this.button('Blackjack enabled', false, 'blackjack enable')} ${this.button('off', true)}`;
-		} else {
-			return `${this.button('Blackjack enabled', true)} ${this.button('off', false, 'blackjack disable')}`;
-		}
-	}
-	language() {
-		if (!this.user.can('editroom', null, this.room)) {
-			return this.button(this.room.language ? Chat.languages.get(this.room.language)! : 'English', true);
-		}
-
-		const languageOutput = [];
-		languageOutput.push(this.button(`English`, !this.room.language, 'roomlanguage english'));
-		for (const [id, text] of Chat.languages) {
-			languageOutput.push(this.button(text, this.room.language === id, `roomlanguage ${id}`));
-		}
-		return languageOutput.join(' ');
-	}
-	generateDisplay() {
-		let output = Chat.html`<div class="infobox">Room Settings for ${this.room.title}<br />`;
-		output += `<strong>Language:</strong> <br />${this.language()}<br />`;
-		output += `<strong>Modchat:</strong> <br />${this.modchat()}<br />`;
-		output += `<strong>Modjoin:</strong> <br />${this.modjoin()}<br />`;
-		output += `<strong>Stretch filter:</strong> <br />${this.stretching()}<br />`;
-		output += `<strong>Caps filter:</strong> <br />${this.capitals()}<br />`;
-		output += `<strong>Emoji filter:</strong> <br />${this.emojis()}<br />`;
-		output += `<strong>Slowchat:</strong> <br />${this.slowchat()}<br />`;
-		output += `<strong>Tournaments:</strong> <br />${this.tourStatus()}<br />`;
-		output += `<strong>UNO:</strong> <br />${this.uno()}<br />`;
-		output += `<strong>Hangman:</strong> <br />${this.hangman()}<br />`;
-		output += `<strong>Blackjack:</strong> <br />${this.blackjack()}<br />`;
-		output += `<strong>Mafia:</strong> <br />${this.mafia()}<br />`;
-		output += '</div>';
-
-		this.user.sendTo(this.room, `|uhtml${(this.sameCommand ? '' : 'change')}|roomsettings|${output}`);
-	}
-}
-
 export const commands: ChatCommands = {
 	roomsetting: 'roomsettings',
 	roomsettings(target, room, user, connection) {
 		if (room.battle) return this.errorReply("This command cannot be used in battle rooms.");
-		const settings = new RoomSettings(user, room, connection);
+		let uhtml = 'uhtml';
 
 		if (!target) {
 			room.update();
-			settings.generateDisplay();
 		} else {
 			this.parse(`/${target}`);
-			settings.updateSetting(target);
+			uhtml = 'uhtmlchange';
+		}
+		// send the settings list
+		function button(setting: string, disable: boolean, command = ' ') {
+			if (disable) {
+				return Chat.html`<button class="button disabled" style="font-weight:bold; color:#575757; font-weight:bold; background-color:#d3d3d3;">${setting}</button> `;
+			}
+			return Chat.html`<button class="button" name="send" value="/roomsetting ${command}">${setting}</button> `;
+		}
+
+		let output = Chat.html`<div class="infobox">Room Settings for ${room.title}<br />`;
+		for (const handler of Chat.roomSettings) {
+			output += handler.call({button}, room, user, connection);
```

Comment:
Hm, I wonder if `{button}` should be replaced with a nicer API that had less code duplication...

It might be useful to nail down API now instead of later.

Reply:
```ts
class SettingsHandler {
  buf: string;
  constructor() {
    this.buf = ``;
  }
  label(text: string) {
    buf += `<strong>${text}:</strong> <br />`;
  }
  button(...etc
```

Reply:
Given that the structure for settings is always the same, what about something like
```ts
type SettingsHandler = (room, user, connection) => {
    label: string, // ie, "Mafia"
    buttons: [string /** button label */, string | null /** button command */][],
}
```

Reply:
something like https://github.com/smogon/pokemon-showdown/compare/master...whalemer:wbi

Reply:
Though, having implemented a few of these, basically everything is either a radio button or an on/off toggle, so maybe we should just have helpers for those.

Reply:
I'm not a huge fan of how that looks in practice. A lot of intermediate variables.

Reply:
```ts
(room, user) => {
	const label = "Language";
	if (!user.can('editroom', null, room)) {
		return {
			label,
			buttons: [[room.language ? Chat.languages.get(room.language)! : 'English', true]],
		};
	}
	const buttons: [string, string | true][] = [];
	buttons.push([`English`, !room.language || 'roomlanguage english']);
	for (const [id, text] of Chat.languages) {
		if (id !== 'english') buttons.push([text, room.language === id || `roomlanguage ${id}`]);
	}
	return {label, buttons};
},
```

vs

```ts
(room, user) => ({
	label: "Language",
	buttons: user.can('editroom', null, room)) ? [
		[`English`, !room.language || 'roomlanguage english'],
		...([...Chat.languages].filter(([id]) => id !== 'english').map(([id, language]) =>
			[language, room.language === id || `roomlanguage ${id}`]
		))
	] : [
		[room.language ? Chat.languages.get(room.language)! : 'English', true]
	],
}),
```

Reply:
I think I would then think about how to design the API to make that look even simpler than it currently is.

I think it would help to just force `Chat.languages` to have English first, so we don't get that messy `...([...Chat.languages])` mess, and then we can just do:

```ts
(room, user) => ({
	label: "Language",
	buttons: (user.can('editroom', null, room)) ?
		[...Chat.languages].map(([id, language]) =>
			[language, room.language === id || `roomlanguage ${id}`]
		)
	:
		[[room.language ? Chat.languages.get(room.language)! : 'English', true]]
	,
}),
```

Reply:
And then we'd just have a `permission` field that, when false, would only show the currently selected option, so it could just be:

```ts
(room, user) => ({
	label: "Language",
	permission: user.can('editroom', null, room),
	buttons: [...Chat.languages].map(([id, language]) =>
		[language, room.language === id || `roomlanguage ${id}`]
	),
}),
```